### PR TITLE
Add Rule Manager to Y-Flow FSMs

### PR DESCRIPTION
**Pull request by dpoltavets**
**MERGE NOTE:** merge after #4632 and change the base correspondingly.

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        Set<PathId> pathIds = flowPaths.stream()
+                .map(FlowPath::getPathId)
+                .collect(Collectors.toSet());
+        DataAdapter dataAdapter = PersistenceDataAdapter.builder()
+                .persistenceManager(persistenceManager)
+                .switchIds(switchIds)
+                .pathIds(pathIds)
+                .build();
+
+        return ruleManager.buildRulesForYFlow(flowPaths, dataAdapter).stream()
```
Comment (sergii-iakovenko) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java:108]: So, it looks like you convert the y-flow into a set of flow paths, and then inside of RM you convert them back. Is it as expected?

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfBatchExecutor.java [hunk 2]
```diff
@@ -85,45 +86,53 @@ public OfBatchExecutor(IOFSwitch iofSwitch, KafkaUtilityService kafkaUtilityServ
...
-        for (String uuid : stageCommandsUuids) {
+        for (UUID uuid : stageCommandsUuids) {
             if (holder.canExecute(uuid)) {
-                BatchData batchData = holder.getByUUid(uuid);
-
-                hasFlows |= batchData.isFlow();
-                hasMeters |= batchData.isMeter();
-                hasGroups |= batchData.isGroup();
-                ofMessages.add(batchData.getMessage());
+                Optional<BatchData> optionalBatchData = holder.getByUUid(uuid);
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfBatchExecutor.java]: Good proposal. Reworked.

File: src-java/floodlight-service/floodlight-api/src/main/java/org/openkilda/floodlight/api/response/rulemanager/SpeakerCommandResponse.java [hunk 3]
```diff
@@ -15,26 +15,40 @@
...
 @Value
-public class SpeakerCommandResponse extends AbstractMessage {
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+public class SpeakerCommandResponse extends SpeakerResponse {
+
+    @JsonProperty("success")
+    boolean success;
+    @JsonProperty("failed_uuids")
+    Map<UUID, String> failedUuids;
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-api/src/main/java/org/openkilda/floodlight/api/response/rulemanager/SpeakerCommandResponse.java]: Renamed.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+public abstract class YFlowSpeakerProcessingAction<T extends YFlowProcessingFsm<T, S, E, C, ?, ?>, S, E, C>
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java]: Renamed.

File: src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/repositories/FermaFlowPathRepository.java [hunk 5]
```diff
@@ -110,6 +111,20 @@ public FermaFlowPathRepository(FermaPersistentImplementation implementation) {
...
     }
 
+    @Override
+    public Map<PathId, YFlow> findYFlowsByPathIds(Set<PathId> pathIds) {
+        Set<String> graphPathIds = pathIds.stream()
+                .map(PathIdConverter.INSTANCE::toGraphProperty)
+                .collect(Collectors.toSet());
+        List<? extends FlowPathFrame> flowPathFrames = framedGraph().traverse(g -> g.V()
+                        .hasLabel(FlowPathFrame.FRAME_LABEL)
+                        .has(FlowPathFrame.PATH_ID_PROPERTY, P.within(graphPathIds)))
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/repositories/FermaFlowPathRepository.java:121]: This method returns `Map<PathId, YFlow>`. Therefore, it is easier to leave `FlowPath`.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfBatchExecutor.java [hunk 6]
```diff
@@ -85,45 +86,53 @@ public OfBatchExecutor(IOFSwitch iofSwitch, KafkaUtilityService kafkaUtilityServ
...
     }
 
-
-    private void checkOfResponses(List<CompletableFuture<Optional<OFMessage>>> waitingMessages) {
-        for (CompletableFuture<Optional<OFMessage>> message : waitingMessages) {
-            try {
-                OFMessage ofMessage = message.get().get();
-                String uuid = holder.popAwaitingXid(ofMessage.getXid());
+    private void checkResponse(Optional<OFMessage> message, Throwable ex, long expectedXid) {
+        if (ex == null) {
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfBatchExecutor.java]: Moved.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java (hunk 1 above)
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java]: At first I did not pick out the paths, but due to the fact that we are using adapters, it turned out that as I did it easier.

**Review by sergii-iakovenko — APPROVED** [review]

**Review by timofei-durakov — APPROVED** [review]

**Pull request merged**